velox/connectors/kafka/KafkaDataSource.cpp

```diff
@@ -0,0 +1,158 @@
+         config_->exists(ConnectionConfig::kGroupId) &&
+         config_->exists(ConnectionConfig::kFormat) && !consumer_.get();
+}
+
+void KafkaDataSource::createConsumer(cppkafka::Configuration& config) {
+  VELOX_CHECK_NULL(
+      consumer_.get(),
+      "Failed to create kafka consumer as the consumer is not null");
+  CppKafkaConsumerPtr cppKafkaConsumer =
+      std::make_shared<cppkafka::Consumer>(config);
+  cppKafkaConsumer->set_destroy_flags(RD_KAFKA_DESTROY_F_NO_CONSUMER_CLOSE);
+  consumer_ = std::make_shared<KafkaConsumer>(
+      cppKafkaConsumer, config_->getPollTimeoutMills(), batchSize_);
+  std::string topic = config_->getTopic();
+  topics_.emplace_back(topic);
+  consumer_->subscribe(topics_);
+}
+
+void KafkaDataSource::createCachedQueue(const uint32_t size) {
+  VELOX_CHECK_GT(
+      size, 0, "Kafka consume message queue size must greater than 0");
+  queue_.reserve(size);
+}
+
+void KafkaDataSource::createRecordDeserializer(
+    const std::string& format,
+    const RowTypePtr& outputType) {
+  if (format == "json") {
+    deserializer_ = std::make_shared<KafkaStreamJSONRecordDeserializer>(
+        outputType, queryCtx_->memoryPool());
+  } else if (format == "csv") {
+    deserializer_ = std::make_shared<KafkaCSVRecordDeserializer>(
+        outputType, queryCtx_->memoryPool());
+  } else if (format == "raw") {
+    deserializer_ = std::make_shared<KafkaRawRecordDeserializer>(
+        outputType, queryCtx_->memoryPool());
+  } else {
+    VELOX_FAIL_UNSUPPORTED_INPUT_UNCATCHABLE(
+        "The data format {} is not supported for kafka.", format);
+  }
+  outRow_ = RowVector::createEmpty(outputType_, queryCtx_->memoryPool());
+  outRow_->resize(1);
+}
+
+void KafkaDataSource::addSplit(ConnectorSplitPtr split) {
+  KafkaConnectorSplit* kafkaConnectorSplit =
+      static_cast<KafkaConnectorSplit*>(split.get());
+  VELOX_CHECK_NOT_NULL(
+      kafkaConnectorSplit,
+      "Failed to add split, because the kafka connector split is null.");
+  VELOX_CHECK_NOT_NULL(
+      consumer_.get(),
+      "Failed to add split, because the kafka consumer is null.");
+  cppkafka::TopicPartitionList topicPartitions =
+      kafkaConnectorSplit->getCppKafkaTopicPartitions();
+  if (topicPartitions.size() == 0) {
+    const auto tps =
+        consumer_->getTopicPartitions(topics_[0], config_->getStartupMode());
+    consumer_->assign(tps);
+  } else {
+    consumer_->setTopicPartitionsOffset(topicPartitions, config_->getStartupMode());
+    consumer_->assign(topicPartitions);
+  }
+}
+
+std::optional<RowVectorPtr> KafkaDataSource::next(
+    uint64_t,
+    velox::ContinueFuture&) {
+  std::optional<RowVectorPtr> res;
+  size_t consumedMsgBytes = 0;
+  if (queue_.empty()) {
+    consumer_->consumeBatch(queue_, consumedMsgBytes);
+    if (consumedMsgBytes == 0) {
+      return res;
+    }
+  }
+  outRow_->prepareForReuse();
+  size_t processDataSize = batchSize_ > 1 ? queue_.size() : batchSize_;
```

Comment:
Here the batchSize seems not work, the real size of output is depend on the consumedMsgBytes.

Reply:
No, when we construct `KafkaConsumer`, we passed the `batchSize` to it, and then the consumer will use it as its `pollBatchSize`, which determine the batch size consumed.  `consumedMsgBytes` here is only to record how many bytes it consumed. @shuai-xu